### PR TITLE
[4.2] Make sure we are not using downloadSources from another version

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue https://github.com/joomla/update.joomla.org/pull/277

### Summary of Changes

Make sure we are not using downloadSources from another version

### Testing Instructions

Install 4.2.6
set this at you update server: "https://update.joomla.org/core/nightlies/next_minor_list.xml"

### Actual result BEFORE applying this Pull Request

Downgrade to 4.0.4

### Expected result AFTER applying this Pull Request

Error message "download failed"

### Link to documentations
Please select:
- [X] No documentation changes for docs.joomla.org needed
- [X] No documentation changes for manual.joomla.org needed

### Analyse

There are two few issues which in combination cause this issue.
- On this XML we have just one item which do use the new downloadsource tag. (4.0.4)
- The download URL for the 4.3.0-dev does trigger an 404 as the version has not been reset yet

Said downloadSources object from 4.0.4 will be carried over as its not overwritten by the 4.3.0-dev object as it does not exists there.

Now the seccond issue comes into play as the original download URL is 404 joomla tries an alternative via the downloadsource  tag. But that tag is set to 4.0.4 URLs which are valid and cause the issue.

Whith the changes here we make sure that this special tag is not carried over to another version when not overriden.

TBH this code itself is far from ideal and should be rewritten for many reasons, but this here should fix the issue as it is for now.

cc @richard67 @obuisard

### CODE REVIEW

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 86d7db176fec5efda5052ea5749e26bdc08ba8a2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39446">issues.joomla.org/tracker/joomla-cms/39446</a>.</sub>

**Comment by richard67** [comment]
In addition to the testing instructions, I have tested with success that the update works as well as without this PR when the update URL exists so the 404 redirect doesn't happen. You can use the following download URL for this test: https://test5.richard-fath.de/next_minor_list.xml

**Comment by zero-24** [comment]
Thanks so one more test here so this can be fixed :)

File: libraries/src/Updater/Update.php [hunk 1]
```diff
@@ -402,6 +402,9 @@ public function _endElement($parser, $name)
             case 'UPDATES':
                 // If the latest item is set then we transfer it to where we want to
                 if (isset($this->latest)) {
+                    // This is an optional tag and therfor we need to be sure that this is gone and only used when set by the update itself
```
Comment (Quy) [inline libraries/src/Updater/Update.php]:
```suggestion
                    // This is an optional tag and therefore we need to be sure that this is gone and only used when set by the update itself
```

**Comment by ReLater** [comment]
Before I confirm as successful (it was).

- You changed https://update.joomla.org/core/nightlies/next_minor_list.xml just for testing this pr? 
- Or will page https://developer.joomla.org/nightly-builds.html provide a new url afterwards?
- Or does this change really mean that i cannot update from 4.2.x to 4.3.0-dev anymore in the future? I always prefer that harakiri way because it's easier to repair a failed update than moving all the content and extensions etc. pp. to the dev version. I always prefer to test with "living content".

**Comment by richard67** [comment]
> Before I confirm as successful (it was).
> 
> * You changed https://update.joomla.org/core/nightlies/next_minor_list.xml just for testing this pr?

No. But the links in the next_minor_extension.xml currently don’t work because of wrong package names.

> * Or will page https://developer.joomla.org/nightly-builds.html provide a new url afterwards?

Yes, these will be changed after that PR had been tested so they have the right package names and the live update will work, too.

> * Or does this change really mean that i cannot update from 4.2.x to 4.3.x-dev anymore in the future?

No. Updating from 4.2.x to 4.3.x-dev shall and will be possible when all is fixed.

File: libraries/src/Updater/Update.php (hunk 1 above)
Comment (richard67) [reply libraries/src/Updater/Update.php]: @zero-24 I've allowed myself to commit this suggestion for you.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 80d67762ed6eb8e36533e63db83041c05e016e2d<br /><br />My previous test is still valid because the change after that was only a fix for a typo in a code comment.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39446">issues.joomla.org/tracker/joomla-cms/39446</a>.</sub>

File: libraries/src/Updater/Update.php (hunk 1 above)
Comment (zero-24) [reply libraries/src/Updater/Update.php]: Thanks

**Comment by zero-24** [comment]
> Before I confirm as successful (it was).

Its broken since the alpha2 commit: https://github.com/joomla/joomla-cms/commit/e1812a9a2a8890f28c2e22be2d228a0b4643d0c9 it should have been -dev only. Or the URL on the update server had to be updated too. Works both ways but not how it is now.

But as Richard said the update server stay the same and we will deploy an workaround for existing sites. Can you confirm this issue and the patch? Than we do have two tests and can deploy the workaround.

**Comment by ReLater** [comment]
I have tested this item :white_check_mark: successfully on 80d67762ed6eb8e36533e63db83041c05e016e2d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39446">issues.joomla.org/tracker/joomla-cms/39446</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39446">issues.joomla.org/tracker/joomla-cms/39446</a>.</sub>

**Comment by ReLater** [comment]
@zero-24 @richard67 Thanks for clarification. I had this issue on 2 sites yesterday but didn't check that version was 4.0.4 after update and grumbled and grumbled while debugging and repairing...

**Comment by zero-24** [comment]
Workaround for the update server prepared here: https://github.com/joomla/update.joomla.org/pull/278#issue-1502072307

**Comment by zero-24** [comment]
Thanks @ReLater took me some debugging too but lucky me i found it :)

**Comment by richard67** [comment]
> Workaround for the update server prepared here: [joomla/update.joomla.org#278 (comment)](https://github.com/joomla/update.joomla.org/pull/278#issue-1502072307)

@zero24 This fixes only the issue with the downgrade to 4.0.4 for sites which don’t have yet the patch of this PR here, but the wrong URL and package name still needs to be fixed, right?

Unfortunately it‘s already late and I need sleep and gotta work tomorrow, so I can’t test that before tomorrow night.

**Comment by zero-24** [comment]
you are right but with that PR the correct message 'download failed" will be shown. The version.php or the version and Download URLs within the xmls needs to be fixed too.

**Comment by roland-d** [comment]
Thank you

**Pull request merged**